transaction/_transaction.py

```diff
@@ -409,6 +416,67 @@ def _callAfterCommitHooks(self, status=True):
         self._after_commit = []
         self._before_commit = []
 
+    def getBeforeAbortHooks(self):
+        """ See ITransaction.
+        """
+        return iter(self._before_abort)
+
+    def addBeforeAbortHook(self, hook, args=(), kws=None):
+        """ See ITransaction.
+        """
+        if kws is None:
+            kws = {}
+        self._before_abort.append((hook, tuple(args), kws))
+
+    def _callBeforeAbortHooks(self):
+        # Call all hooks registered, allowing further registrations
+        # during processing.  Note that calls to addBeforeAbortHook() may
+        # add additional hooks while hooks are running, and iterating over a
+        # growing list is well-defined in Python.
+        for hook, args, kws in self._before_abort:
```

Comment:
I wonder if calling the before abort hooks should be wrapped in `try/except:` handling.

I know the before *commit* hooks aren't wrapped in that, but that's ok: if they throw, it's possible to catch that and abort the transaction; one con do this to programatically implement dooming transactions, for example:
```python
try:
   tx.commit()
except DoomedTransaction:
  tx.abort()
```

So it's OK if a before-commit hook raises. But if a before abort hook raises, then we're left with no way to actually abort the transaction.

Reply:
Changed as suggested.